pkg/controller/support/credentials/cloudauth/gcr/gcr_test.go

```diff
@@ -1,102 +1,321 @@
+			gcrRegistryErr,
+		},
+		{
+			"failed-do-request",
+			"gcr.io",
+			canceledCtx,
+			nil,
+			&gcrProvider{
+				logger:      log,
+				tokenSource: &fakeOauth2TokenSource{},
+			},
+			nil,
+			cloudauthtest.FakeChallengeLoginServer(
+				"serviceName",
+				ts.URL,
+				nil),
+			ctxCanceledErr.Error(),
+			ctxCanceledErr,
+		},
+		{
+			"non-200-response-code",
+			"gcr.io",
+			defaultCtx,
+			roundTripFunc(func(r *http.Request) (*http.Response, error) {
+				return &http.Response{
+					StatusCode: http.StatusBadRequest,
+					Body:       io.NopCloser(strings.NewReader(`nope`)),
+				}, nil
+			}),
+			&gcrProvider{
+				logger:      log,
+				tokenSource: &fakeOauth2TokenSource{},
+			},
+			nil,
+			cloudauthtest.FakeChallengeLoginServer(
+				"serviceName",
+				ts.URL,
+				nil),
+			non200StatusErr.Error(),
+			non200StatusErr,
+		},
+		{
+			"failed-no-token-in-response",
+			"gcr.io",
+			defaultCtx,
+			roundTripFunc(func(r *http.Request) (*http.Response, error) {
+				res, err := json.Marshal(tokenResponse{})
 				if err != nil {
-					t.Fatalf("%#v", err)
+					t.Fatalf("Unexpected error while marshalling token response")
 				}
-				if auth.Username == "" || auth.Password == "" || auth.RegistryToken == "" {
-					t.Fatalf("incorrect auth config: %v", auth)
+				return &http.Response{
+					StatusCode: http.StatusOK,
+					Body:       io.NopCloser(strings.NewReader(string(res))),
+				}, nil
+			}),
+			&gcrProvider{
+				logger:      log,
+				tokenSource: &fakeOauth2TokenSource{},
+			},
+			nil,
+			cloudauthtest.FakeChallengeLoginServer(
+				"serviceName",
+				ts.URL,
+				nil),
+			noResTokenErr.Error(),
+			noResTokenErr,
+		},
+		// success
+		{
+			"token-response-has-access-token",
+			"gcr.io",
+			defaultCtx,
+			roundTripFunc(func(r *http.Request) (*http.Response, error) {
+				res, err := json.Marshal(tokenResponse{
+					AccessToken: "test-access-token",
+				})
+				if err != nil {
+					t.Fatalf("Unexpected error while marshalling token response")
 				}
-
-				// verify the registry token
-				req, err := http.NewRequestWithContext(ctx, "GET", "https://"+tt.name+"/v2/", nil)
+				return &http.Response{
+					StatusCode: http.StatusOK,
+					Body:       io.NopCloser(strings.NewReader(string(res))),
+				}, nil
+			}),
+			&gcrProvider{
+				logger:      log,
+				tokenSource: &fakeOauth2TokenSource{},
+			},
+			&types.AuthConfig{
+				Username:      "oauth2accesstoken",
+				Password:      "hey",
+				RegistryToken: "test-access-token",
+			},
+			cloudauthtest.FakeChallengeLoginServer(
+				"serviceName",
+				ts.URL,
+				nil),
+			"Successfully authenticated with gcr \"gcr.io\"",
+			nil,
+		},
+		{
+			"token-response",
+			"gcr.io",
+			defaultCtx,
+			roundTripFunc(func(r *http.Request) (*http.Response, error) {
+				res, err := json.Marshal(tokenResponse{
+					Token: "regular-token",
+				})
 				if err != nil {
-					t.Fatal(err)
+					t.Fatalf("Unexpected error while marshalling token response")
```

Comment:
surface the error as well

Reply:
👍🏽 I decided to move this out into a helper fn since there was a lot of repeat code in creating the round tripper. LMK what ya think

https://github.com/dominodatalab/hephaestus/pull/67/files#diff-5a64af87db87fed02f741d838e0256f07ac212ebbc0eea7269d969b6046cdb3cR240